### PR TITLE
[SwiftUI] `AvatarGroup` SwiftUI API

### PR DESCRIPTION
**Pull request by mischreiber**
### Platforms Impacted
- [x] iOS
- [x] visionOS
- [ ] macOS

### Description of changes

Creating a SwiftUI API for `AvatarGroup`.

Yes, it's always been a SwiftUI control, but the API was only ever designed to be used in a UIKit wrapper, where `Avatar` elements could be added and removed via event-driven method calls. Our new API is fully data-driven, as befits a native SwiftUI control.

```swift
let style = MSFAvatarGroupStyle.stack
let size = MSFAvatarSize.size40
let avatarCount = 5

// Basic initializer--there are more properties to explore!
AvatarGroup(style: style,
            size: size,
            avatarCount: avatarCount) { index in

    // Create and yield an Avatar for the given callback index.
    Avatar(style: style,
           size: size,
           image: fetchImage(index),
           primaryText: fetchName(index))
}
```

As part of this change, I've also avoided the ability to mark an `AvatarGroup` as transparent, aligning with our latest designs. This will eventually be deprecated across all existing `Avatar` endpoints.

### Binary change

Total increase: 12,544 bytes
Total decrease: -52,432 bytes
| File | Before | After | Delta |
|------|-------:|------:|------:|
| Total | 31,203,440 bytes | 31,163,552 bytes | 🎉 -39,888 bytes |
<details>
<summary> Full breakdown </summary>

| File | Before | After | Delta |
|------|-------:|------:|------:|
| CalendarViewWeekdayHeadingView.o | 77,928 bytes | 83,480 bytes | ⚠️ 5,552 bytes |
| SegmentedControl.o | 330,216 bytes | 333,640 bytes | ⚠️ 3,424 bytes |
| PopupMenuController.o | 375,464 bytes | 378,064 bytes | ⚠️ 2,600 bytes |
| MSFAvatarGroup.o | 31,184 bytes | 32,152 bytes | ⚠️ 968 bytes |
| BottomSheetController.o | 525,872 bytes | 525,512 bytes | 🎉 -360 bytes |
| MultilineCommandBar.o | 140,104 bytes | 139,352 bytes | 🎉 -752 bytes |
| IndeterminateProgressBar.o | 231,928 bytes | 231,112 bytes | 🎉 -816 bytes |
| HUD.o | 267,176 bytes | 266,200 bytes | 🎉 -976 bytes |
| Avatar.o | 604,704 bytes | 596,968 bytes | 🎉 -7,736 bytes |
| __.SYMDEF | 4,863,752 bytes | 4,855,520 bytes | 🎉 -8,232 bytes |
| AvatarGroup.o | 417,456 bytes | 383,896 bytes | 🎉 -33,560 bytes |
</details>

### Verification

<details>
<summary>Visual Verification</summary>
SwiftUI

https://github.com/microsoft/fluentui-apple/assets/4934719/e2fb5aa0-23d2-4e17-b27c-9ce1e92fbbd1

### UIKit

https://github.com/microsoft/fluentui-apple/assets/4934719/c5f694fb-bd74-4942-b73e-c9b2f9a2863c

</details>

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [x] iOS supported versions (all major versions greater than or equal current target deployment version)
- [x] VoiceOver and Keyboard Accessibility
- [x] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [x] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/fluentui-apple/pull/2022)

### CODE REVIEW

File: ios/FluentUI/AvatarGroup/AvatarGroup.swift [hunk 1]
```diff
@@ -149,57 +172,33 @@ public struct AvatarGroup: View, TokenizedControlView {
...
 
             // Calculating the size delta of the current and next avatar based off of ring visibility, which helps determine
             // starting coordinates for the cutout.
-            let currentAvatarHasRing = avatar.isRingVisible
-            let nextAvatarHasRing = !isLastDisplayed ? avatars[nextIndex].isRingVisible : false
+            let currentAvatarHasRing = avatar.state.isRingVisible
+            let nextAvatarHasRing = !isLastDisplayed ? avatars[nextIndex].state.isRingVisible : false
 
             // Calculating the different interspace scenarios considering rings.
-            let ringOuterGap = avatarView.tokenSet[.ringOuterGap].float
```
Comment (sophialee0416) [inline ios/FluentUI/AvatarGroup/AvatarGroup.swift:167]: Hooray for code cleanup!

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/AvatarGroupDemoController_SwiftUI.swift [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                            overflowCount: overflowCount,
+                            isUnread: isUnread) { index in
+                    avatarFromSamplePersona(index)
+                }.fixedSize()
+                
+            }
+            .frame(height: 120, alignment: .center)
+
+            FluentList {
+                Section("Content") {
```
Comment (alexanderboswell) [inline ios/FluentUI.Demo/FluentUI.Demo/Demos/AvatarGroupDemoController_SwiftUI.swift:76]: can use `FluentSection` now with `FluentSectionHeader` :)

File: ios/FluentUI/AvatarGroup/AvatarGroup.swift [hunk 3]
```diff
@@ -95,24 +95,47 @@ public struct AvatarGroup: View, TokenizedControlView {
...
+    ///   - maxDisplayedAvatars: Caps the number of displayed avatars and shows the remaining not displayed in the overflow avatar.
+    ///   - overflowCount: Adds to the overflow count in case the calling code did not provide all the avatars, but still wants to convey more
+    ///                    items than just the remainder of the avatars that could not be displayed due to the maxDisplayedAvatars property.
+    ///   - isUnread: Show a top-trailing aligned unread dot when set to true.
+    ///   - avatarBuilder: A ViewBuilder that creates an `Avatar` for a given index.
     public init(style: MSFAvatarGroupStyle,
-                size: MSFAvatarSize) {
+                size: MSFAvatarSize,
+                avatarCount: Int = 0,
+                maxDisplayedAvatars: Int = 5,
```
Comment (alexanderboswell) [inline ios/FluentUI/AvatarGroup/AvatarGroup.swift:111]: I assume this is non-zero so we always show something if a implementation doesn't specify it, and 5 is just an arbitrary value picked?

File: ios/FluentUI/AvatarGroup/AvatarGroup.swift [hunk 4]
```diff
@@ -266,8 +265,8 @@ public struct AvatarGroup: View, TokenizedControlView {
     var displayedAvatarAccessibilityLabels: [String] {
         var labels: [String] = []
         for i in 0..<avatarsToDisplay {
-            let avatar = state.avatars[i]
-            labels.append(avatar.accessibilityLabel ?? avatar.primaryText ?? avatar.secondaryText ?? "")
+            let avatarState = state.avatars[i].state
+            labels.append(avatarState.accessibilityLabel ?? avatarState.primaryText ?? avatarState.secondaryText ?? "")
```
Comment (alexanderboswell) [inline ios/FluentUI/AvatarGroup/AvatarGroup.swift:269]:
```swift
if let label = avatarState.accessibilityLabel ?? avatarState.primaryText ?? avatarState.secondaryText {
labels.append(label)
}
```

Is there a need to append a empty string?

**Review by alexanderboswell — APPROVED** [review]

**Review by nodes11 — APPROVED** [review]

File: ios/FluentUI/AvatarGroup/AvatarGroup.swift (hunk 3 above)
Comment (mischreiber) [reply ios/FluentUI/AvatarGroup/AvatarGroup.swift:111]: Good catch--that was originally just for testing. The default should be `Int.max`.

File: ios/FluentUI/AvatarGroup/AvatarGroup.swift (hunk 4 above)
Comment (mischreiber) [reply ios/FluentUI/AvatarGroup/AvatarGroup.swift:269]: I agree--not a fan of having that blank in the array of accessibility items.

**Pull request merged**